HPXMLtoOpenStudio/resources/hvac.rb

```diff
@@ -5411,6 +5417,102 @@ def self.apply_advanced_defrost(model, htg_coil, air_loop_unitary, conditioned_s
     )
   end
 
+  # Creates an EMS program to add pan heater energy use for a heat pump.
+  # A pan heater ensures that water melted during the defrost cycle does not refreeze into ice and
+  # result in fan obstruction or coil damage.
+  #
+  # @param model [OpenStudio::Model::Model] OpenStudio Model object
+  # @param htg_coil [OpenStudio::Model::CoilHeatingDXSingleSpeed or OpenStudio::Model::CoilHeatingDXMultiSpeed] OpenStudio Heating Coil object
+  # @param air_loop_unitary [OpenStudio::Model::AirLoopHVACUnitarySystem] Air loop for the HVAC system
+  # @param conditioned_space [OpenStudio::Model::Space] OpenStudio Space object for conditioned zone
+  # @param heat_pump [HPXML::HeatPump] The HPXML heat pump of interest
+  # @param heating_unavailable_periods [HPXML::UnavailablePeriods] Unavailable periods for heating
+  # @return [nil]
+  def self.apply_pan_heater_ems_program(model, htg_coil, air_loop_unitary, conditioned_space, heat_pump, heating_unavailable_periods)
+    # Other equipment/actuator
+    cnt = model.getOtherEquipments.count { |e| e.endUseSubcategory.start_with? Constants::ObjectTypePanHeater } # Ensure unique meter for each heat pump
+    pan_heater_energy_oe = Model.add_other_equipment(
+      model,
+      name: "#{air_loop_unitary.name} pan heater energy",
+      end_use: "#{Constants::ObjectTypePanHeater}#{cnt + 1}",
+      space: conditioned_space,
+      design_level: 0,
+      frac_radiant: 0,
+      frac_latent: 0,
+      frac_lost: 1,
+      schedule: model.alwaysOnDiscreteSchedule,
+      fuel_type: HPXML::FuelTypeElectricity
+    )
+    pan_heater_energy_oe.additionalProperties.setFeature('HPXML_ID', heat_pump.id) # Used by reporting measure
+
+    pan_heater_energy_oe_act = Model.add_ems_actuator(
+      name: "#{pan_heater_energy_oe.name} act",
+      model_object: pan_heater_energy_oe,
+      comp_type_and_control: EPlus::EMSActuatorOtherEquipmentPower
+    )
+
+    # Sensors
+    tout_db_sensor = Model.add_ems_sensor(
+      model,
+      name: "#{air_loop_unitary.name} tout s",
+      output_var_or_meter_name: 'Site Outdoor Air Drybulb Temperature',
+      key_name: 'Environment'
+    )
+
+    if heat_pump.pan_heater_control_type == HPXML::HVACPanHeaterControlTypeDefrost
+      htg_coil_rtf_sensor = Model.add_ems_sensor(
+        model,
+        name: "#{htg_coil.name} rtf s",
+        output_var_or_meter_name: 'Heating Coil Runtime Fraction',
+        key_name: htg_coil.name
+      )
+    end
+
+    # Create HVAC availability sensor
+    if not heating_unavailable_periods.empty?
+      htg_avail_sch = ScheduleConstant.new(model, 'heating availability schedule', 1.0, EPlus::ScheduleTypeLimitsFraction, unavailable_periods: heating_unavailable_periods)
```

Comment:
Not super familiar with unavailable periods, my understanding is that we need it here because we don't want to calculate the pan heater energy use in `continuous` mode if the hvac system is unavailable and outdoor temperature is below 32F. When `defrost mode`, the coil runtime fraction will be 0.0 during unavailable periods, so for advanced defrost EMS, we don't need to worry about the unavailable period, right? This approach looks good to me, just want to double check the advanced defrost is working well.

Reply:
Yes, that exactly matches my thinking. I assume this is only needed for `continuous` mode option. But I will manually test a HPXML w/ the advanced defrost model and a power outage to double-check it's not needed there.

Reply:
Confirmed that it looks good, the defrost supplemental heating energy was zero during the power outage.